flang/lib/Lower/OpenACC.cpp

```diff
@@ -119,160 +119,233 @@ static Op createSimpleOp(Fortran::lower::FirOpBuilder &builder,
   return op;
 }
 
-static void genACC(Fortran::lower::AbstractConverter &converter,
-                   Fortran::lower::pft::Evaluation &eval,
-                   const Fortran::parser::OpenACCLoopConstruct &loopConstruct) {
-
-  const auto &beginLoopDirective =
-      std::get<Fortran::parser::AccBeginLoopDirective>(loopConstruct.t);
-  const auto &loopDirective =
-      std::get<Fortran::parser::AccLoopDirective>(beginLoopDirective.t);
+static void genAsyncClause(Fortran::lower::AbstractConverter &converter,
+                           const Fortran::parser::AccClause::Async *asyncClause,
+                           mlir::Value &async, bool &addAsyncAttr) {
+  const auto &asyncClauseValue = asyncClause->v;
+  if (asyncClauseValue) { // async has a value.
+    async = fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(*asyncClauseValue)));
+  } else {
+    addAsyncAttr = true;
+  }
+}
 
-  if (loopDirective.v == llvm::acc::ACCD_loop) {
+static void genDeviceTypeClause(
+    Fortran::lower::AbstractConverter &converter,
+    const Fortran::parser::AccClause::DeviceType *deviceTypeClause,
+    SmallVectorImpl<mlir::Value> &operands) {
+  const auto &deviceTypeValue = deviceTypeClause->v;
+  if (deviceTypeValue) {
+    for (const auto &scalarIntExpr : *deviceTypeValue) {
+      mlir::Value expr = fir::getBase(
+          converter.genExprValue(*Fortran::semantics::GetExpr(scalarIntExpr)));
+      operands.push_back(expr);
+    }
+  } else {
     auto &firOpBuilder = converter.getFirOpBuilder();
-    auto currentLocation = converter.getCurrentLocation();
+    // * was passed as value and will be represented as a -1 constant
+    // integer.
+    mlir::Value star = firOpBuilder.createIntegerConstant(
+        converter.getCurrentLocation(), firOpBuilder.getIntegerType(32),
```

Comment:
Should this type be hard-coded like this?

Reply:
Do you mean the `-1` or the whole Value creation?

Reply:
Just the -1 (for constexpr). But if the IR pattern is fairly common, you might want a function.

Reply:
Since this is just how we represent the `*` passed in the clause it cannot be changed by the user. Of course we could decide to represent that with a i64 if we would like.

Reply:
I'll switch to `index`. The op is accepting `index` as well.